### ISSUE TITLE
Connection error when shipping UDF

### ISSUE THREAD

**Issue by DonDebonair** [cloudera/impyla#45]
Hi!

I finally had some time to try out the Python UDF goodness, but I ran into some problems. I'm running a 3-node CDH5.1.2 cluster on Amazon AWS. I tried one of the simple UDFs that is used in the Impyla tests, like this:

``` python
from impala.udf import udf, ship_udf
from impala.udf.types import (FunctionContext, BooleanVal, SmallIntVal, IntVal,
                              BigIntVal, StringVal)
from impala.context import ImpalaContext

@udf(BigIntVal(FunctionContext, IntVal))
def int_promotion(context, x):
    return x + 1

ctx = ImpalaContext(nn_host='ec2-54-171-167-66.eu-west-1.compute.amazonaws.com', host='ec2-54-171-176-150.eu-west-1.compute.amazonaws.com', hdfs_user='hdfs', webhdfs_port=50070)

ship_udf(ctx, int_promotion, overwrite=True)
```

which gives me the following error:

```
ConnectionError: HTTPConnectionPool(host='ip-172-31-16-101.eu-west-1.compute.internal', port=50075): Max retries exceeded with url: /webhdfs/v1/tmp/impyla-MLXIPYFS/int_promotion.ll?op=CREATE&user.name=hdfs&namenoderpcaddress=ip-172-31-16-100.eu-west-1.compute.internal:8020&overwrite=true (Caused by <class 'socket.gaierror'>: [Errno 8] nodename nor servname provided, or not known)
```

I'm sure it's something simple like opening the right port on either the NN or the Data Nodes, but I'm not seeing it :)

**Comment by DandyDev** [comment]
Just wanted to let you know that I solved my problem! 

**TLDR:** Amazon is trying to be smart with DNS resolving so webhdfs can't access the datanode. Hacking `/etc/hosts` helps.

Longer explanation for future reference:

It has to do with the way Amazon handles DNS resolving.  Each EC2 instance gets an internal IP address and a public IP, and coupled to those are an internal domain name (`ip-xx-xx-xx-xx.eu-west-1.compute.internal`) for the internal IP and an external domain name (`ec2-xx-xx-xx-xx.eu-west-1.compute.amazonaws.com`) for the public IP.

Amazon tries to be smart, so when you use the external domain name on one of your EC2 instances, it will resolve to the _internal_ IP, and when you use that same external domain name outside of Amazon, it will resolve to the public IP.

The problem was that, although I used external domain names when installing the Hadoop cluster, internally those would still resolve to the internal IPs. So when the UDF gets shipped to the cluster, _webhdfs_ will first ask the namenode (for which I provided the external domain name through the impala context) for the applicable datanode to store the blocks, and it will get back an internal IP which resolve to an _internal_ domain name. This cannot be accessed from the outside, so webhdfs cannot save the file to HDFS.

I have sadly not found a way to solve this on the cluster side, so for now I just put the internal domain names in my `/etc/hosts` locally, with the corresponding public IPs. Now I can ship UDFs, and it's awesome!

**Comment by andreisavu** [comment]
Hi Daan! I'm glad you found  a solution to this problem. Some alternatives you may want to consider later on that would work properly without tricky DNS settings are:

* run the UDF script from a gateway node within the cluster and reference the internal hostnames for various services. It may be annoying to move files back and forth but it could be simplified by mounting a filesystem over ssh. 

* setup a VPN / DirectConnect link. This is a great option if this is a setup you will often use for development / production work. It basically extends your local network with an AWS subnet and allows seamless connectivity over the private interfaces.

**Comment by DandyDev** [comment]
Hi Andrei! Thanks for the tips! We'll probably go for the first option. We already have a server running on Digital Ocean, which is doing various Hadoop related things, and we were planning to move it to Amazon anyways. If we stick it in the same VPC, we can use it for pushing UDFs and the likes as well :)
